test/e2e/suite/issuers/acme/certificate/notafter.go

```diff
@@ -0,0 +1,153 @@
+			})
+		Expect(err).NotTo(HaveOccurred())
+		By("Verifying the ACME account URI is set")
+		err = util.WaitForIssuerStatusFunc(f.CertManagerClientSet.CertmanagerV1().Issuers(f.Namespace.Name),
+			issuerName,
+			func(i *v1.Issuer) (bool, error) {
+				if i.GetStatus().ACMEStatus().URI == "" {
+					return false, nil
+				}
+				return true, nil
+			})
+		Expect(err).NotTo(HaveOccurred())
+		By("Verifying ACME account private key exists")
+		secret, err := f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Get(context.TODO(), testingACMEPrivateKey, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		if len(secret.Data) != 1 {
+			Fail("Expected 1 key in ACME account private key secret, but there was %d", len(secret.Data))
+		}
+	})
+
+	JustBeforeEach(func() {
+		acmeIngressDomain = frameworkutil.RandomSubdomain(f.Config.Addons.IngressController.Domain)
+	})
+
+	AfterEach(func() {
+		By("Cleaning up")
+		f.CertManagerClientSet.CertmanagerV1().Issuers(f.Namespace.Name).Delete(context.TODO(), issuerName, metav1.DeleteOptions{})
+		f.KubeClientSet.CoreV1().Secrets(f.Namespace.Name).Delete(context.TODO(), testingACMEPrivateKey, metav1.DeleteOptions{})
+	})
+
+	It("should obtain a signed certificate with a single CN from the ACME server with 1 hour validity", func() {
+		certClient := f.CertManagerClientSet.CertmanagerV1().Certificates(f.Namespace.Name)
+
+		By("Creating a Certificate")
+		cert := gen.Certificate(certificateName,
+			gen.SetCertificateDuration(time.Hour),
+			gen.SetCertificateRenewBefore(45*time.Minute),
+			gen.SetCertificateSecretName(certificateSecretName),
+			gen.SetCertificateIssuer(cmmeta.ObjectReference{Name: issuerName}),
+			gen.SetCertificateDNSNames(acmeIngressDomain),
+		)
+		cert.Namespace = f.Namespace.Name
+
+		_, err := certClient.Create(context.TODO(), cert, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for the Certificate to be issued...")
+		err = f.Helper().WaitCertificateIssued(f.Namespace.Name, certificateName, time.Minute*5)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Validating the issued Certificate...")
+		err = f.Helper().ValidateCertificate(f.Namespace.Name, certificateName)
+		Expect(err).NotTo(HaveOccurred())
+
+		sec, err := f.Helper().WaitForSecretCertificateData(f.Namespace.Name, certificateSecretName, time.Minute*5)
+		Expect(err).NotTo(HaveOccurred(), "failed to wait for secret")
+
+		crtPEM := sec.Data[corev1.TLSCertKey]
+		crt, err := pki.DecodeX509CertificateBytes(crtPEM)
+		Expect(err).NotTo(HaveOccurred(), "failed to get decode signed certificate data")
+
+		// checking loosely to not hit too many timing issues as the date is defined in the controller
+		// pebble issues a 5 year cert by default
+		if crt.NotAfter.After(time.Now().Add(time.Hour)) {
+			Fail(fmt.Sprintf("Certificate has a NotAfter time after more than 1 hour (requested duration), got %s, current time %s", crt.NotAfter.String(), time.Now().String()))
+		}
```

Comment:
Looks good.

I wondered if Pebble also allows you to set a NotBefore field, in which case it'd be possible to check the actual  duration (NotAfter - NotBefore)  of the signed certificate exactly matched the Certificate.Spec.Duration.

Reply:
it does but our k8s API doesn't really lean towards a not before function adding that means supporting setting that field